packages/teuchos/parameterlist/src/Teuchos_VerboseObjectParameterListHelpers.cpp

```diff
@@ -117,19 +137,94 @@ void Teuchos::readVerboseObjectSublist(
+    //      can give the illusion that things are working, if each
+    //      process writes exactly the same information (e.g., solver
+    //      convergence information for a bulk synchronous solve).
+    //      This introduces a terrible scalability problem, as the
+    //      filesystem is then tasked with coping with concurrent writes
+    //      to the same shared file, which is should make you cry a little.
+    //
+    // The resolution, is two fold:
+    //   1st, construct the ostream as a regular wrapper around cout
+    //   2nd, restrict the file creation and opening to a single process
+    //   Finally, map all ostreams except the fstream one to
+    //    a blackhole. Ensuring each rank has a functional stream
+    //    but that only one actually emits data to disk
+    //
+    // A second optimization is added, where the user can control the buffer
+    // sized used by the ofstream. It was not immediately obvious where this
+    // buffer could live, so a static std::vector of std::vector<char> has
+    // been added to this source file. By using an STL object, we can esnure
+    // that the buffers are destructed at program exit. Another option would
+    // be to modify FancyOStream to store this buffer.
+    //
+    *oStream = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+    #if defined(HAVE_TEUCHOS_MPI)
+    const auto rank = Teuchos::GlobalMPISession::getRank();
+    #else
+    const auto rank = outputFileMPIRank;
+    #endif
+    if ( rank  == outputFileMPIRank) {
+      RCP<std::ofstream> oFileStream = rcp(new std::ofstream());
+      // to allow for user configurable buffer sizes, we need to construct
+      // the fstream in two steps
+      if (outputFileBufferSize != OutputFile_buffer_size_kb_default) {
+        // create a dummy object, that we push onto our array of vectors
+        std::vector<char> dummy(0);
+        OutputFile_buffers_.push_back(dummy);
+        // grab the object stored in the vector
+        std::vector<char>& buff = OutputFile_buffers_.back();
+        // allocate the buffer 
+        buff.reserve(outputFileBufferSize*1024);
+        buff.resize(outputFileBufferSize*1024);
+        // attach the buffer to the stream
+        oFileStream->rdbuf()->pubsetbuf(buff.data(), buff.size());
+      }
+      // open the fstream only on a single MPI process
+      oFileStream->open(outputFileStr);
+  
+      TEUCHOS_TEST_FOR_EXCEPTION_PURE_MSG(
+        oFileStream->eof(), Exceptions::InvalidParameterValue,
+        "Error, the file \"" << outputFileStr << "\n given by the parameter\n"
+        "\'" << OutputFile_name << "\' in the sublist\n"
+        "\'" << voSublist.name() << "\' count not be opened for output!"
+        );
+      // wrap the fstream inside fancyOStream
+      *oStream = fancyOStream(rcp_implicit_cast<std::ostream>(oFileStream));
+    }
+
+    #if defined(HAVE_TEUCHOS_MPI)
+    // ensure that only one stream actually emits data
```

Comment:
Doesn't this render ineffective all the changes above?

Reply:
No, that step is necessary. If you follow the logic, the initial OStream, is actually std::cout. The ensures each process has an ostream.

The final tidbit, simply blackholes everyone that didn't go through the if condition (the rank allowed output is the rank that goes into the if.

Unless, I botched the guards...